sourcegraph/sourcegraph
# Re-write Cody Web Client

Sorry for the big PR, but there was no other way. 

This PR almost entirely re-writes the Cody web client and state management. The old state management was powered by zustand and not native react states. It then internally integrated the chat client, which was non-reactive. The old chat client was just builder function, and it was required to re-create the client with every single change. The old client then also used callbacks to update the state at the parent level, and the state was partially duplicated. Basically, it needed a makeover.

- The new client is reactive and uses native react states.
- The new chat store allows having separate chat states for the sidebar and standalone chat.
- The states are not partially duplicated at two places anymore.
- It also prepares for multi-repo scope.
- It introduces URL-based routing for /chat.
- It disables the editor widget when a message is in progress.
- It also prepares for editing any message and not just the last one.
- It also fixes the re-rendering of the whole chat on input change.

Future PRs:
- update Codebase context and context fetcher clients to support new CodyClientScope aka multi-repo context
- implement cody scope selector UI
- save the scope with each interaction and the current scope with the transcript.
- load the current scope from the transcript on load.
- show scope with each human message.
- show scope selector with edit message form.
- Allow editing of any message and not just the last.

## Test plan

- visit /cody on web and check if messages, history and all other actions are working as expected.
- Check same for vs code & app.
